oximeter/oximeter/src/schema.rs

```diff
@@ -0,0 +1,640 @@
+        self.timeseries_name == other.timeseries_name
+            && self.datum_type == other.datum_type
+            && self.field_schema == other.field_schema
+    }
+}
+
+// Regular expression describing valid timeseries names.
+//
+// Names are derived from the names of the Rust structs for the target and metric, converted to
+// snake case. So the names must be valid identifiers, and generally:
+//
+//  - Start with lowercase a-z
+//  - Any number of alphanumerics
+//  - Zero or more of the above, delimited by '-'.
+//
+// That describes the target/metric name, and the timeseries is two of those, joined with ':'.
+const TIMESERIES_NAME_REGEX: &str =
+    "(([a-z]+[a-z0-9]*)(_([a-z0-9]+))*):(([a-z]+[a-z0-9]*)(_([a-z0-9]+))*)";
+
+/// A set of timeseries schema, useful for testing changes to targets or
+/// metrics.
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct SchemaSet {
+    #[serde(flatten)]
+    inner: BTreeMap<TimeseriesName, TimeseriesSchema>,
+}
+
+impl SchemaSet {
+    /// Insert a timeseries schema, checking for conflicts.
+    ///
+    /// This inserts the schema derived from `target` and `metric`. If one
+    /// does _not_ already exist in `self` or a _matching_ one exists, `None`
+    /// is returned.
+    ///
+    /// If the derived schema _conflicts_ with one in `self`, the existing
+    /// schema is returned.
+    pub fn insert_checked<T, M>(
+        &mut self,
+        target: &T,
+        metric: &M,
+    ) -> Option<TimeseriesSchema>
+    where
+        T: Target,
+        M: Metric,
+    {
+        let new = TimeseriesSchema::new(target, metric);
+        let name = new.timeseries_name.clone();
+        match self.inner.entry(name) {
+            Entry::Vacant(entry) => {
+                entry.insert(new);
+                None
+            }
+            Entry::Occupied(entry) => {
+                let existing = entry.get();
+                if existing == &new {
+                    None
+                } else {
+                    Some(existing.clone())
+                }
+            }
+        }
+    }
+
+    /// Compare the set of schema against the contents of a file.
+    ///
+    /// This function loads a `SchemaSet` from the provided JSON file, and
+    /// asserts that the contained schema matches those in `self`. Note that
+    /// equality of `TimeseriesSchema` ignores creation timestamps, so this
+    /// compares the "identity" data: timeseries name, field names, field types,
+    /// and field sources.
+    ///
+    /// This is intentionally similar to `expectorate::assert_contents()`. If
+    /// the provided file doesn't exist, it's treated as empty. If it does, a
+    /// `SchemaSet` is deserialized from it and a comparison between that and
+    /// `self` is done.
+    ///
+    /// You can use `EXPECTORATE=overwrite` to overwrite the existing file,
+    /// rather than panicking.
+    pub fn assert_contents(&self, path: impl AsRef<Path>) {
```

Comment:
This is the meat of the PR, the function that loads a checked-in schema filed and asserts it matches the ones in `self`. Note that this is a bit different from `expectorate` itself, since we're checking equality in a way that ignores the creation timestamp of the schema.